### PR TITLE
Optimize java generic signature creation

### PR DESCRIPTION
**Pull request by retronym**
References scala/scala-dev#418.

I haven't changed the logic to use `$` rather than `.`, but I
think we should clarify the intent of this and compare with
javac's behaviour.

### CODE REVIEW

**Comment by retronym** [comment]
The `jardiff` after bootstrapping through this change is empty, as expected.

[Benchmark](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?var-scalaSha=a0026c5%7C7c30f9d5&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40&from=1474264982264&to=1503555511298) scheduled.

**Comment by retronym** [comment]
A theory about why scalap showed the biggest speedup: it has some generic signatures involving lots of nested type constructors, which might have started to suffer from exponential-time performance in the old implementation.

```scala
trait Rule[-In, +Out, +A, +X] extends (In => Result[Out, A, X]) {
   ...
  /** ^~~~~~^(f) is equivalent to ^^ { case b1 ~ b2 ~ b3 ~ b4 ~ b5 ~ b6 => f(b1, b2, b3, b4, b5, b6) }
   */
  def ^~~~~~^[B1, B2, B3, B4, B5, B6, B >: A <% B1 ~ B2 ~ B3 ~ B4 ~ B5 ~ B6, C](f: (B1, B2, B3, B4, B5, B6) => C) = map { a =>
    (a: B1 ~ B2 ~ B3 ~ B4 ~ B5 ~ B6) match { case b1 ~ b2 ~ b3 ~ b4 ~ b5 ~ b6 => f(b1, b2, b3, b4, b5, b6) }
  }

  /** ^~~~~~~^(f) is equivalent to ^^ { case b1 ~ b2 ~ b3 ~ b4 ~ b5 ~ b6 => f(b1, b2, b3, b4, b5, b6) }
   */
  def ^~~~~~~^[B1, B2, B3, B4, B5, B6, B7, B >: A <% B1 ~ B2 ~ B3 ~ B4 ~ B5 ~ B6 ~ B7, C](f: (B1, B2, B3, B4, B5, B6, B7) => C) = map { a =>
    (a: B1 ~ B2 ~ B3 ~ B4 ~ B5 ~ B6 ~ B7) match { case b1 ~ b2 ~ b3 ~ b4 ~ b5 ~ b6 ~b7 => f(b1, b2, b3, b4, b5, b6, b7) }
  }
```

```
⚡ jardiff -c build/quick/classes/scalap/ | grep 'signature '| awk '{ print length, $0 }' | sort -nr -s | cut -d" " -f2- | head
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B7:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function7<TB1;TB2;TB3;TB4;TB5;TB6;TB7;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;TB7;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn2;TOut2;TC;TX2;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B7:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function7<TB1;TB2;TB3;TB4;TB5;TB6;TB7;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;TB7;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn2;TOut2;TC;TX2;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B7:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function7<TB1;TB2;TB3;TB4;TB5;TB6;TB7;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;TB7;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn;TOut;TC;TX;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B7:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function7<TB1;TB2;TB3;TB4;TB5;TB6;TB7;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;TB7;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn;TOut;TC;TX;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B7:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function7<TB1;TB2;TB3;TB4;TB5;TB6;TB7;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;TB7;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn;TOut;TC;TX;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function6<TB1;TB2;TB3;TB4;TB5;TB6;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn2;TOut2;TC;TX2;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function6<TB1;TB2;TB3;TB4;TB5;TB6;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn2;TOut2;TC;TX2;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function6<TB1;TB2;TB3;TB4;TB5;TB6;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn;TOut;TC;TX;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function6<TB1;TB2;TB3;TB4;TB5;TB6;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn;TOut;TC;TX;>;
+  // signature <B1:Ljava/lang/Object;B2:Ljava/lang/Object;B3:Ljava/lang/Object;B4:Ljava/lang/Object;B5:Ljava/lang/Object;B6:Ljava/lang/Object;B:Ljava/lang/Object;C:Ljava/lang/Object;>(Lscala/Function6<TB1;TB2;TB3;TB4;TB5;TB6;TC;>;Lscala/Function1<TB;Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<Lscala/tools/scalap/scalax/rules/$tilde<TB1;TB2;>;TB3;>;TB4;>;TB5;>;TB6;>;>;)Lscala/tools/scalap/scalax/rules/Rule<TIn;TOut;TC;TX;>;
```

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 1]
```diff
@@ -236,82 +215,108 @@ abstract class Erasure extends InfoTransform
...
-
-      (ps map boxedSig).mkString
+      ps.foreach(boxedSig)
     }
-    def boxedSig(tp: Type) = jsig(tp, primitiveOK = false)
-    def boundsSig(bounds: List[Type]) = {
+    def boxedSig(tp: Type): Unit = jsig(tp, primitiveOK = false)
+    def boundsSig(bounds: List[Type]): Unit = {
       val (isTrait, isClass) = bounds partition (_.typeSymbol.isTrait)
       val classPart = isClass match {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: `classPart` is unused - and storing its value doesn't add any information over its compile-time type :-)

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 2]
```diff
@@ -236,82 +215,108 @@ abstract class Erasure extends InfoTransform
...
+          def classSig: Unit = {
             markClassUsed(sym)
             val preRebound = pre.baseType(sym.owner) // #2585
-            val sigCls = {
-              if (needsJavaSig(preRebound, Nil)) {
-                val s = jsig(preRebound, existentiallyBound)
-                if (s.charAt(0) == 'L') {
-                  val withoutSemi = s.substring(0, s.length - 1)
+
+              val sigCls = {if (needsJavaSig(preRebound, Nil)) {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: `sigCls` is also unused. also strange formatting for the opening `{`.

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 3]
```diff
@@ -236,82 +215,108 @@ abstract class Erasure extends InfoTransform
...
+                  // This arises due to multiply nested classes in the face of the
+                  // rewriting explained at rebindInnerClass.
+
+                  // TODO revisit this. Does it align with javac for code that can be expressed in both languages?
+                  val delimiter = if (builder.charAt(builder.length() - 1) == '>') '.' else '$'
+                  builder.append(delimiter).append(sym.javaSimpleName)
                 } else fullNameInSig(sym)
               }
               else fullNameInSig(sym)
+}
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: indentation

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 4]
```diff
@@ -236,82 +215,108 @@ abstract class Erasure extends InfoTransform
...
-              if (needsJavaSig(preRebound, Nil)) {
-                val s = jsig(preRebound, existentiallyBound)
-                if (s.charAt(0) == 'L') {
-                  val withoutSemi = s.substring(0, s.length - 1)
+
+              val sigCls = {if (needsJavaSig(preRebound, Nil)) {
+                val i = builder.length()
+                jsig(preRebound, existentiallyBound)
+                if (builder.charAt(i) == 'L') {
+                   val withoutSemi =builder.delete(builder.length() - 1, builder.length())
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: `withoutSemi` is unused (and anyway an alias for `builder`). could use `deleteCharAt(builder.length())`

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 5]
```diff
@@ -88,31 +88,6 @@ abstract class Erasure extends InfoTransform
...
-      if (ch == '.' && last != '>') {
-         copy(i) = '$'
-         changed = true
-      }
-      last = ch
-      i += 1
-    }
-    if (changed) new String(copy) else sig
-  }
-
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:115]: 👍 👍

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 6]
```diff
@@ -236,82 +215,108 @@ abstract class Erasure extends InfoTransform
...
 
     // Anything which could conceivably be a module (i.e. isn't known to be
     // a type parameter or similar) must go through here or the signature is
     // likely to end up with Foo<T>.Empty where it needs Foo<T>.Empty$.
-    def fullNameInSig(sym: Symbol) = "L" + enteringJVM(sym.javaBinaryNameString)
+    def fullNameInSig(sym: Symbol): Unit = builder.append('L').append(enteringJVM(sym.javaBinaryNameString))
 
-    def jsig(tp0: Type, existentiallyBound: List[Symbol] = Nil, toplevel: Boolean = false, primitiveOK: Boolean = true): String = {
+    @noinline
+    def jsig(tp0: Type, existentiallyBound: List[Symbol] = Nil, toplevel: Boolean = false, primitiveOK: Boolean = true, prefixMode: Boolean = false): Unit = {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: The new `prefixMode` parameter seems to be unused - or maybe I missed its usage?

**Review by lrytz — COMMENTED** [review]
Only some minor cleanups, LGTM otherwise

**Comment by retronym** [comment]
Fixed the one comment that applied to the final commit after squashing the commits together which addressed the other formatting and unused val problems.

**Review by lrytz — APPROVED** [review]

**Pull request merged**